versioned_docs/version-2021sp/lecture3.md

```diff
@@ -43,4 +43,358 @@ From now on, we will be using a `tsconfig.json` file within every Node project w
 
 Use the link above to install Postman.
 
-The rest of the notes is coming soon!
+## Postman and Request Bodies
+
+### Postman
+
+Instead of always going to the endpoint in the browser, a robust way of testing
+our endpoints is to use Postman.
+
+Postman is a software that allows you to simulate requests that could be sent by
+a user to your backend. It is useful for testing and ensuring that the behavior
+of your requests (including necessary headers) is what you expect.
+
+### POST Request
+
+Usually when you want to send a `POST` request you also want to send information
+with it. Situationally, you want to do this using request bodies rather than
+query parameters.
+
+Let's say we have this `addSong` endpoint:
+
+```ts title="index.ts"
+app.post('/addSong', (req, res) => {
+  const song: songtype = { name: req.body.name, rating: req.body.rating };
+  console.log(song);
+  songs.push(song);
+  res.send(`Song ${req.body.name} added!`);
+});
+```
+
+where `songtype` is:
+
+```ts title="index.ts"
+type songtype = {
+  name: string;
+  rating: number;
+};
+```
+
+Previously, we may have considered using query parameters for sending data for
+the backend. There's nothing wrong with that; for example, we could have used
+`/addSong?name=Despacito&rating=5`. However, this can lead to extremely long
+URLs, and limit us from sending more complicated data. That's where **request
+bodies** come in handy. We can instead send request data in JSON format to the
+backend, allowing us to use the data more easily and integrate it seamlessly
+with our backend (which happens to be in TypeScript, so we can easily deal with
+it).
+
+The snippet above tells `express` to listen for `POST` requests at endpoint
+`/addSong`. `req.body` is a JavaScript object, and we access its properties
+`req.body.name` and `req.body.rating` to add a new song to our array of songs.
+
+Before this snippet actually works, we need to import something called
+`body-parser`. This lets our express server to actually read `req.body`
```

Comment:
Maybe instruct them to also `yarn add` it as a devdependency

Reply:
The code snippet right below has a comment above the `import` line telling them to `yarn add body-parser`.

Also it's just a regular dependency, right?

Reply:
oops you're right about both nvm